docs/program-setup/bare_metal/mbed2_porting.md

```diff
@@ -0,0 +1,109 @@
+
+```
+"supported_c_libs": {
+    "arm":  ["std"], ["small"],
+    "gcc_arm":  ["std"], ["small"],
+    "iar": ["std"]
+}
+```
+<span class="notes">**Note:** For ARM, your target scatter file needs to have a two-region memory model. If the build system throws an error when compiling with microlib, you can find more information [here](https://os.mbed.com/docs/mbed-os/v6.0-preview/reference/using-small-c-libraries.html).</span>
+
+<span class="notes">**Note:** The IAR toolchain does not have a small C library.</span>
+
+- Set the `c_lib` property to indicate which C library should be linked by the build tools. If your target has `"default_lib": "small"` defined, then replace with `"c_lib" : "small"`. Otherwise add `"c_lib" : "small"`. We recommend that ultra constrained devices running bare metal applications link with the small C libraries by default.
+
+<span class="notes">**Note:** If a toolchain does not support a small C library and `c_lib` is set to `small`, the build tools will revert to linking with the standard C library, provided that it is listed in `supported_c_libs` for the given toolchain.</span>
+
+- If `default_toolchain` is set to `uARM`, then replace it with `ARM` and remove `uARM` from `supported_toolchains`. Support for the uARM toolchain, which is the ARM toolchain with microlib, has been removed. Setting `c_lib` to `small` ensures that microlib is used when the ARM toolchain is selected.
+
+<span class="notes">**Note:** As mentioned above, to successfully build with microlib, the target must define a two-region memory model. You might need to replace the content of the TOOLCHAIN_ARM_STD linker file with that of TOOLCHAIN_ARM_MICRO which already includes a two-region memory model linker file.
+
+you might need to replace the content of the TOOLCHAIN_ARM_STD scatter file with that of TOOLCHAIN_ARM_MICRO. This is because the TOOLCHAIN_ARM_MICRO linker file defines a two-region memory model which is necessary to build with microlib.</span>
+
+- If your board does not have a low power ticker, ensure that tickless is enabled using the microsecond ticker as follows:
+```
+"overrides": {
+    "tickless-from-us-ticker": true
+}
+```
+
+- It might be necessary to reduce the stack size allocated for your target if it does not have enough RAM. The stack size is configured by setting a value for the `boot-stack-size` attribute; this value must be a multiple of 8 for alignment purposes. By default all targets are configured to have a boot stack size of 0x1000 (4096 bytes) in bare metal. However, this must be overridden if inadequate for your target. We recommend to reduce the boot stack size to 0x400 (1096 bytes) if your target has 8KB of RAM and to 0x300 (768 bytes) if your target has 4KB of RAM.
```

Comment:
0x400 is 1024 bytes

Reply:
I'm not convinced, as far as I remember bare metal has only one stack. So the boot stack is the thread stack as well and the ISR stack.

> 0x400 (1096 bytes) if your target has 8KB of RAM 
7K for global data and heap seems excessive. Also we'd like to have the applications pretty much portable if the stacks are different between different boards. The initial idea was that boards won't be allowed to play with stack sizes, only applications will be allowed to change them.

> 0x300 (768 bytes) if your target has 4KB of RAM.

Is it even possible to run Mbed on 4k board? Considering it's probably 8-16k flash...

Reply:
> I'm not convinced, as far as I remember bare metal has only one stack. So the boot stack is the thread stack as well and the ISR stack.
> 
> > 0x400 (1096 bytes) if your target has 8KB of RAM
> > 7K for global data and heap seems excessive. Also we'd like to have the applications pretty much portable if the stacks are different between different boards. The initial idea was that boards won't be allowed to play with stack sizes, only applications will be allowed to change them.
> 
> > 0x300 (768 bytes) if your target has 4KB of RAM.
> 
> Is it even possible to run Mbed on 4k board? Considering it's probably 8-16k flash...

You are right that in bare metal the boot stack is the thread stack as well and the ISR stack. And that stack is defined to be 4K by default in https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json/#L60. For ultra constrained devices, it needs to be lowered.

LPC1114FN28: Flash 32kB, RAM 4kB was ported to bare metal https://github.com/ARMmbed/mbed-os/pull/12623/commits/90d7e6266fe2c582f0b938bf4444dec7f425f723;  the recommendation to set `boot-stack-size` to `0x300` comes from that port.

Reply:
OK, but this is an extreme example with target that small it doesn't make much sense to run Mbed and it makes even less sense for us to spend time on making it work. I'm afraid that this exceptions will lead to a mess when people see "oh you can do that as it's already done for this target". It would be more than reasonable to say that we require at least 32/48K of ROM and at 8K of RAM. How big is bare metal blinky these days?